src/Redistribution.sol

```diff
@@ -400,119 +396,80 @@ contract Redistribution is AccessControl, Pausable {
-                // where probability is stakeDensity / currentSum
-                // to avoid resorting to floating points all divisions should be
-                // simplified with multiplying both sides (as long as divisor > 0)
-                // randomNumber / (MaxH + 1) < stakeDensity / currentSum
-                // ( randomNumber / (MaxH + 1) ) * currentSum < stakeDensity
-                // randomNumber * currentSum < stakeDensity * (MaxH + 1)
-                if (randomNumberTrunc * currentSum < currentReveals[revIndex].stakeDensity * (uint256(MaxH) + 1)) {
-                    truthRevealedHash = currentReveals[revIndex].hash;
-                    truthRevealedDepth = currentReveals[revIndex].depth;
-                }
-            }
+        // Freeze those bastards
+        for (uint256 i = 0; i < _frozenOverlays.length; i++) {
+            Stakes.freezeDeposit(
+                _frozenOverlays[i],
+                penaltyMultiplierDisagreement * roundLength * uint256(2 ** truthRevealedDepth)
+            );
         }
 
-        emit TruthSelected(truthRevealedHash, truthRevealedDepth);
-
-        k = 0;
-
-        string memory winnerSelectionAnchor = currentWinnerSelectionAnchor();
-
-        for (uint256 i = 0; i < currentCommits.length; i++) {
-            revIndex = currentCommits[i].revealIndex;
-            if (currentCommits[i].revealed) {
-                if (
-                    truthRevealedHash == currentReveals[revIndex].hash &&
-                    truthRevealedDepth == currentReveals[revIndex].depth
-                ) {
-                    currentWinnerSelectionSum += currentReveals[revIndex].stakeDensity;
-                    randomNumber = keccak256(abi.encodePacked(winnerSelectionAnchor, k));
-
-                    randomNumberTrunc = uint256(randomNumber & MaxH);
-
-                    if (
-                        randomNumberTrunc * currentWinnerSelectionSum <
-                        currentReveals[revIndex].stakeDensity * (uint256(MaxH) + 1)
-                    ) {
-                        winner_ = currentReveals[revIndex];
-                    }
-
-                    k++;
-                } else {
-                    Stakes.freezeDeposit(
-                        currentReveals[revIndex].overlay,
-                        penaltyMultiplierDisagreement * roundLength * uint256(2 ** truthRevealedDepth)
-                    );
-                    // slash ph5
-                }
-            } else {
-                // slash in later phase
-                // Stakes.slashDeposit(currentCommits[i].overlay, currentCommits[i].stake);
-                Stakes.freezeDeposit(
-                    currentCommits[i].overlay,
-                    penaltyMultiplierNonRevealed * roundLength * uint256(2 ** truthRevealedDepth)
-                );
-                continue;
-            }
+        // Slash those other bastards
+        for (uint256 i = 0; i < _slashedOverlays.length; i++) {
```

Comment:
the length currently is 32, won't it cause error when it wants to freeze the a zeroaddress overlay? (e.g. on 20th index there is no value)

Reply:
It won't, as in slashing or freezing there in the staking contract there is this part that checks

`if (stakes[overlay].isValue) {`

which will be false for 0x00 addy, so it will just be skipped without any problems.
Ideal this would be a dynamic array but in this case it's not possible to have dynamic memory arrays, we could use dynamic storage arrays but that would be an even bigger overhead.